### ISSUE TITLE
Details for validation failures not shown with recursion

### ISSUE THREAD

**Issue by tschaub** [stac-utils/stac-validator#192]
It looks like validating with `--recursive -1` doesn't show details for invalid items (and perhaps other types).

Here is an example catalog with a link to an item that does not validate:

`tests/test_data/v100/catalog-with-bad-item.json`
```json
{
  "id": "examples",
  "type": "Catalog",
  "title": "Example Catalog",
  "stac_version": "1.0.0",
  "description": "This is a valid catalog with links to one valid item and one invalid item.",
  "links": [
    {
      "rel": "root",
      "href": "./catalog.json",
      "type": "application/json"
    },
    {
      "rel": "item",
      "href": "./bad-item.json",
      "type": "application/json",
      "title": "Invalid item"
    },
    {
      "rel": "item",
      "href": "./collectionless-item.json",
      "type": "application/json",
      "title": "Collection with no items (standalone)"
    },
    {
      "rel": "self",
      "href": "https://raw.githubusercontent.com/radiantearth/stac-spec/v1.0.0/examples/catalog.json",
      "type": "application/json"
    }
  ]
}
```

And the bad item (missing `id`):
`tests/test_data/v100/bad-item.json`
```json
{
  "stac_version": "1.0.0",
  "stac_extensions": [
    "https://stac-extensions.github.io/eo/v1.0.0/schema.json"
  ],
  "type": "Feature",
  "bbox": [
    -122.59750209,
    37.48803556,
    -122.2880486,
    37.613537207
  ],
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          -122.308150179,
          37.488035566
        ],
        [
          -122.597502109,
          37.538869539
        ],
        [
          -122.576687533,
          37.613537207
        ],
        [
          -122.2880486,
          37.562818007
        ],
        [
          -122.308150179,
          37.488035566
        ]
      ]
    ]
  },
  "properties": {
    "title": "Full Item",
    "description": "A STAC item without an id",
    "datetime": null,
    "start_datetime": "2016-05-03T13:22:30Z",
    "end_datetime": "2016-05-03T13:27:30Z",
    "created": "2016-05-04T00:00:01Z",
    "updated": "2017-01-01T00:30:55Z",
    "license": "various",
    "providers": [
      {
        "name": "Remote Data, Inc",
        "description": "Producers of awesome spatiotemporal assets",
        "roles": [
          "producer",
          "processor"
        ],
        "url": "http://remotedata.it"
      }
    ],
    "platform": "cool_sat2",
    "instruments": [
      "cool_sensor_v1"
    ]
  },
  "links": [
    {
      "rel": "root",
      "href": "./catalog.json",
      "type": "application/json",
      "title": "Example Catalog"
    },
    {
      "rel": "parent",
      "href": "./catalog.json",
      "type": "application/json",
      "title": "Example Catalog"
    },
    {
      "rel": "alternate",
      "type": "text/html",
      "href": "http://cool-sat.com/catalog/CS3-20160503_132130_04/CS3-20160503_132130_04.html",
      "title": "HTML representation of this STAC Item"
    },
    {
      "rel": "license",
      "type": "text/html",
      "href": "http://remotedata.io/license.html",
      "title": "Data License for Remote Data, Inc."
    }
  ],
  "assets": {
    "analytic": {
      "href": "http://cool-sat.com/catalog/CS3-20160503_132130_04/analytic.tif",
      "title": "4-Band Analytic",
      "eo:bands": [
        {
          "name": "band1"
        },
        {
          "name": "band1"
        },
        {
          "name": "band2"
        },
        {
          "name": "band3"
        }
      ]
    }
  }
}
```

Running `stac_validator` against this `catalog-with-bad-item.json`:

```bash
> stac_validator tests/test_data/v100/catalog-with-bad-item.json --recursive -1
[
    {
        "version": "1.0.0",
        "path": "tests/test_data/v100/catalog-with-bad-item.json",
        "schema": [
            "https://schemas.stacspec.org/v1.0.0/catalog-spec/json-schema/catalog.json"
        ],
        "valid_stac": true,
        "asset_type": "CATALOG",
        "validation_method": "recursive"
    }
]
> echo $?
0
```

From the output and the exit code, it looks like things are valid.  Instead, I would expect to see details about the invalid item and a non-zero exit code.

There are two separate issues.

One is that the main function only exits with non-zero status if `recursive == -2`: https://github.com/sparkgeo/stac-validator/blob/e10e90075e54f58c9a318b1dc7f2d56c42f5b1fb/stac_validator/stac_validator.py#L86-L87

See #193 for a proposed fix for the above issue.

The other issue is that during recursive validation, messages are only appended if items are valid.  Since validation exceptions stop recursion, it seems like it would make sense to unconditionally append the message here (after it has been augmented by the exception handler): https://github.com/sparkgeo/stac-validator/blob/e10e90075e54f58c9a318b1dc7f2d56c42f5b1fb/stac_validator/validate.py#L314-L315

**Comment by jonhealy1** [comment]
@tschaub this makes sense:
```The other issue is that during recursive validation, messages are only appended if items are valid. Since validation exceptions stop recursion, it seems like it would make sense to unconditionally append the message here (after it has been augmented by the exception handler):``` 

If you want to have a go at fixing it please do - thank you

**Comment by tschaub** [comment]
> If you want to have a go at fixing it please do - thank you

Back at a keyboard late Monday.  I’ll take a look then.

**Comment by tschaub** [comment]
Proposed fix in #202.
